### DISCUSSION TITLE
opendds for android

### DISCUSSION THREAD

**Discussion by ChandanGirish** [OpenDDS/OpenDDS#4221] [General]
Hello Community,

I am new to opendds and I have few questions regarding integrating opendds on android architecture.

1. Is opendds implemented on application level or on system level of android architecture? Does it use the android native library and HAL, OR it implements a custom vendor HAL and libraries?

2. Can I use RTPS stack to build it at system level in an android source project and use abstraction interface to communicate at app level? because I see there are no android.mk or android.bp files to be built as android source code.

It would be really helpful for me if I could find answers for these questions. Thank you!

**Comment by iguessthislldo** [comment]
On Android OpenDDS mostly uses normal C/Unix APIs. The only Android-specific things it can use are `__android_log_print` and `NetworkCallback`, but I think both of those are optional. I don't know of a reason why it couldn't be used by something in Android itself as long as it's not the kernel, however I'm not familiar with lower level parts of Android and I've only run OpenDDS as part of a normal app or as command line applications. It doesn't support Android-specific build systems, so you'd probably have to build it separately using [theses instructions](https://opendds.readthedocs.io/en/latest-release/building/android.html).

**Reply by iguessthislldo** [reply]
If you need an example of using OpenDDS in a normal application, we have one [here](https://github.com/OpenDDS/opendds-smart-lock).

**Comment by ChandanGirish** [comment]
Thank you
